Add a `flatten` argument to `ssh.libs`

ValekoZ opened the pull request:
This option makes us able to avoid getting the file tree of the remote server and just downloads the desired files in the output folder:

```python
>>> shell = ssh(host='example.org')
>>> shell.libs('/bin/sh', directory = 'libs', flatten = True)
{'/path/to/cwd/libs/libc.so.6': 22708292161536, '/path/to/cwd/libs/ld-linux-x86-64.so.2': 22708294139904, '/path/to/cwd/libs/dash': 0}
```

ValekoZ committed:
Add a `flatten` argument to `ssh.libs`

This option makes us able to avoid getting the file tree of the remote
server and just downloads the desired files in the output folder

ValekoZ committed:
Update changelog

peace-maker commented:
We should emit a warning on duplicate files when the filename is equal but the remote path was different.

ValekoZ commented:
Do you think it should just emit a warning and skip the download ? Or should it renames the file to something like `{filename}_duplicate` and then `{filename}_duplicate2`, etc. ?

peace-maker commented:
I guess renaming them is better instead of discarding used libraries, but I'd switch to the unflattened structure anyways if this happens. We should log a warning in any case. I don't know how stable the order of the loaded libraries is in ldd and if you can be sure you always want the first matching library on every run.

ValekoZ committed:
Add a warning in ssh.libs for duplicate filenames

peace-maker committed:
Merge branch 'dev' into dev

peace-maker commented:
Can you document that fallback behaviour in the argument description please?

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1756,20 +1756,39 @@ def unlink(self, file):
+            for lib in libs:
+                name = os.path.basename(lib)
+
+                if name in basenames:
```
This seems wrong, `in` checks if the value exists in the keys of the dictionary iirc.

ValekoZ replied:
Oh shit sorry, I was using a list at the beginning and then switched to dict in order to avoid to recompute the basename :(
I will fix that

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1756,20 +1756,39 @@ def unlink(self, file):
+                name = os.path.basename(lib)
+
+                if name in basenames:
+                    self.warning('Duplicate lib name, switching to unflattened file tree: %r' % name)
```
Can you output both conflicting remote paths here?

ValekoZ committed:
Add documentation in ssh.libs for duplicate filenames fallback

ValekoZ commented:
Yep, I just did it, sorry I forgot that

peace-maker reviewed (changes requested):
Cool, it would be awesome if you could add tests for the function testing with`flatten` on and off and with and without duplicates.

We're using doctests, check out other examples in the function docs comments of other functions in the file!

ValekoZ committed:
[ssh.libs] Fix flatten's duplicate check

Fix this: https://github.com/Gallopsled/pwntools/pull/2268#discussion_r1325404026

ValekoZ committed:
[ssh.libs] Add better warning for flatten's duplicates

Fix this: https://github.com/Gallopsled/pwntools/pull/2268#discussion_r1325405234

ValekoZ commented:
I just tried to create unit tests for this functionnality, but I was wondering what design did you like better ?
I have the choice between creating minimal ELF files dynamically using lief for creating a binary with 2 libs with the same filename, or just dropping the binaries I created in the repo so that I don't have to add a dependency for the tests

peace-maker commented:
I think adding another small binary to the `pwnlib/data/elf` folder seems fair. It doesn't have to be runnable I guess, so referencing duplicate not-existing libraries might be fine.

ValekoZ committed:
[ssh.libs] Add doctests

This commit adds:

- The `no_duplicate` binary that depends on
    - `a/lib.so`
    - `b/lib2.so`
- The `duplicate` binary that depends on
    - `a/lib.so`
    - `b/lib.so`

Then, the doctest tries to pull the libs from both binaries with
different `flatten` values.

ValekoZ committed:
[ssh.libs] Add `remote` arg to documentation

ValekoZ commented:
Can you check that I added the tests correctly ? It says all tests passed but I'm not even sure the tests were run .. 😅

peace-maker commented:
I think it's saner to abort on duplicate libraries instead of silently falling back to unflattened output. The locations would be different than what the script expects and it would break anyways.
> but I'd switch to the unflattened structure anyways if this happens

I was talking about me using this feature, not how I'd imagine it to behave automatically. Sorry for the confusion.

You can test if the tests are run by adding something obviously broken and see if it breaks.

ValekoZ committed:
[ssh.libs] Remove fallback to unflattened when flatten fails

Arusekk committed:
Merge branch 'dev' into dev

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1804,20 +1804,53 @@ def unlink(self, file):
+
+        Examples:
+            >>> s =  ssh(host='example.pwnme')
+            >>> s.cwd = f'{pwnlib.data.elf.path}/ssh_libs'
```
Please create an `__init__.py` file and use `pwnlib.data.elf.ssh_libs.get("duplicate")` etc. below. Check the the `__init__.py` in one of the other sub folders. The cwd has to be writeable and it apparently isn't in CI, when you change to the pwnlib data directory.

peace-maker reviewed (changes requested):
Nice, the CI ran now and revealed some problems with your tests. Can you `strip` the binaries to reduce size please?

ValekoZ committed:
[ssh.libs] Fix doctests

ValekoZ committed:
[ssh.libs] Import pwnlib.data.elf.ssh_libs in pwnlib.data.elf

ValekoZ commented:
Apparently, the elf I pushed are corrupted (I can't run `ldd` on them), I'll try to fix that ^^

ValekoZ committed:
[ssh.libs] Fix corrupted binaries

ValekoZ committed:
[ssh.libs] We actually need to set the cwd since libs paths are relative

ValekoZ committed:
[ssh.libs] Fix permission issues for CI

ValekoZ committed:
[ssh.libs] Fix doctest

ValekoZ commented:
Sorry for all the commits, I can't run the tests on my computer (it doesn't work, I don't know why)
I still have an issue and I don't understand why, I stop for today I'll check that later ^^

peace-maker commented:
@Arusekk can you have a look please?

peace-maker committed:
Remove doctest attempts

peace-maker committed:
Merge branch 'dev' into dev

peace-maker commented:
Bleh, I've tried to get the tests to work nicely but don't want to wrestle with -rpath $ORIGIN and friends anymore. I'll add this now and deal with the test again later.
